fix(e2e): MLflow /version endpoint returns 401 in RC Release Validation (Kind E2E)

## Summary

The `test_mlflow_version_endpoint` test fails in the **RC Release Validation** workflow (Kind E2E job) with a 401 response, while it passes in the regular `E2E K8s 1.35.0 (Kind)` workflow.

**Failing run:** https://github.com/kagenti/kagenti/actions/runs/26776833102
**Passing run (main, same day):** https://github.com/kagenti/kagenti/actions/runs/26786187730

## Error

```
AssertionError: MLflow /version endpoint failed: 401 - {"detail":"Authentication required"}
assert 401 in (200, 302, 307)
```

Test: `kagenti/tests/e2e/common/test_mlflow_auth.py::TestMLflowAuth::test_mlflow_version_endpoint`

## Root Cause Analysis

### 1. `/version` is not in the unprotected routes list

The `mlflow-oidc-auth` plugin's `_is_unprotected_route()` only exempts `/v1/traces` (via the runtime patch in `charts/kagenti-deps/templates/mlflow.yaml` lines 206-211). The `/version` endpoint is not exempted, so it requires auth when OIDC is enabled.

### 2. 401 vs 302 depends on request type

The `mlflow-oidc-auth` plugin (v7.x) distinguishes between browser requests (returns 302 redirect to OIDC login) and API requests (returns 401 JSON). The test uses `httpx.AsyncClient` without `Accept: text/html`, so the plugin treats it as an API client and returns 401 instead of redirecting.

### 3. `mlflow-oidc-auth` is unpinned

```dockerfile
pip install --no-cache-dir --quiet "mlflow-oidc-auth"
```

This installs whatever version is latest (currently 7.3.1). The v7.2.0 release added "re-authenticate sessions against IdP token expiry" which changed redirect/auth-failure semantics. Without a version pin, behavior can change between runs.

### 4. K8s probes use `/version` but aren't affected

The readiness/liveness probes at lines 413-426 of `mlflow.yaml` use `/version` and work because they hit the pod directly (kubelet → pod IP, not through the auth middleware path). But external requests go through the full middleware stack.

## Proposed Fix

**Option A (recommended): Add `/version` to unprotected routes**

In `charts/kagenti-deps/templates/mlflow.yaml`, extend the runtime patch:

```python
def patched_is_unprotected(self, path: str) -> bool:
    if path.startswith("/v1/traces") or path in ("/version", "/health", "/ping"):
        return True
    return original_is_unprotected(self, path)
```

This is the correct behavior — health/version endpoints should never require auth.

**Option B (also recommended): Pin `mlflow-oidc-auth` version**

```dockerfile
pip install --no-cache-dir --quiet "mlflow-oidc-auth==7.3.1"
```

Prevents surprise behavior changes from upstream releases.

## Context

- Both workflows use `scripts/kind/setup-kagenti.sh --with-all --build-images`
- `--with-all` sets `mlflow.auth.enabled=true` via Helm `--set` (overrides `dev_values.yaml`)
- The mlflow-oauth-secret job completes successfully (OIDC client + K8s secret created)
- Test results in failing run: 1 failed, 71 passed, 30 skipped

## Files to Modify

- `charts/kagenti-deps/templates/mlflow.yaml` (unprotected routes patch, ~line 206)
- Optionally: `kagenti/tests/e2e/common/test_mlflow_auth.py` (accept 401 as valid when auth is enabled)

Assisted-By: Claude Code

--- a/charts/kagenti-deps/templates/mlflow.yaml
+++ b/charts/kagenti-deps/templates/mlflow.yaml
@@ -192,24 +192,41 @@ spec:
           # IMPORTANT: Patch BEFORE importing app (app is created at import time)
           from mlflow_oidc_auth.middleware.auth_middleware import AuthMiddleware
           import mlflow_oidc_auth.middleware.fastapi_permission_middleware as fpm
+          import mlflow_oidc_auth.hooks.before_request as _before_request
 
           # Assert patched APIs exist — fail fast on mlflow-oidc-auth upgrades
           assert hasattr(AuthMiddleware, "_is_unprotected_route"), \
               "mlflow-oidc-auth API changed: AuthMiddleware._is_unprotected_route not found"
           assert hasattr(fpm, "_find_fastapi_validator"), \
               "mlflow-oidc-auth API changed: _find_fastapi_validator not found"
+          assert hasattr(_before_request, "_is_unprotected_route"), \
+              "mlflow-oidc-auth API changed: before_request._is_unprotected_route not found"
 
-          # Patch 1: Exclude /v1/traces from OIDC auth
+          # Patch 1: Exclude /v1/traces, /version, /ping from OIDC auth (FastAPI gate)
           # The OTel collector authenticates via OAuth2 client credentials (Bearer token)
           # but may not have a user profile in the MLflow users table.
-          # Skip auth entirely — Istio mTLS protects this path.
+          # /version and /ping are health-check endpoints that should never require auth.
+          # Skip auth entirely — Istio mTLS protects /v1/traces.
           original_is_unprotected = AuthMiddleware._is_unprotected_route
           def patched_is_unprotected(self, path: str) -> bool:
-              if path.startswith("/v1/traces") or path in ("/version", "/health", "/ping"):
+              if path.startswith("/v1/traces") or path in ("/version", "/ping"):
                   return True
               return original_is_unprotected(self, path)
           AuthMiddleware._is_unprotected_route = patched_is_unprotected
-          print("Excluded /v1/traces, /version, /health, /ping from OIDC auth")
+          print("Excluded /v1/traces, /version, /ping from OIDC auth (FastAPI gate)")
+
+          # Patch 1b: Exclude /version, /ping from Flask before_request auth gate
+          # /version and /ping are Flask routes that pass through both FastAPI middleware
+          # AND the Flask before_request hook. Without this patch, Gate A (above) skips
+          # auth (so no username is set), then Gate B returns 401 because username is None.
+          # /health is already natively exempt at this layer.
+          _original_flask_is_unprotected = _before_request._is_unprotected_route
+          def _patched_flask_is_unprotected(path: str) -> bool:
+              if path in ("/version", "/ping"):
+                  return True
+              return _original_flask_is_unprotected(path)
+          _before_request._is_unprotected_route = _patched_flask_is_unprotected
+          print("Excluded /version, /ping from Flask before_request auth gate")
 
           # Patch 2: Skip permission validation for /v1/traces
           # mlflow-oidc-auth v7+ adds a FastAPI permission middleware that checks

--- a/charts/kagenti-deps/templates/mlflow.yaml
+++ b/charts/kagenti-deps/templates/mlflow.yaml
@@ -131,7 +131,7 @@ spec:
           export HOME=/tmp && \
           pip install --no-cache-dir --quiet "psycopg[binary]" && \
           {{- if .Values.mlflow.auth.enabled }}
-          pip install --no-cache-dir --quiet "mlflow-oidc-auth" && \
+          pip install --no-cache-dir --quiet "mlflow-oidc-auth>=7.3.1,<8" && \
           {{- end }}
           {{- if and .Values.openshift .Values.mlflow.auth.enabled }}
           # Combine system CAs with the ingress CA so Python can verify both
@@ -205,11 +205,11 @@ spec:
           # Skip auth entirely — Istio mTLS protects this path.
           original_is_unprotected = AuthMiddleware._is_unprotected_route
           def patched_is_unprotected(self, path: str) -> bool:
-              if path.startswith("/v1/traces"):
+              if path.startswith("/v1/traces") or path in ("/version", "/health", "/ping"):
                   return True
               return original_is_unprotected(self, path)
           AuthMiddleware._is_unprotected_route = patched_is_unprotected
-          print("Excluded /v1/traces from OIDC auth (mTLS via Istio)")
+          print("Excluded /v1/traces, /version, /health, /ping from OIDC auth")
 
           # Patch 2: Skip permission validation for /v1/traces
           # mlflow-oidc-auth v7+ adds a FastAPI permission middleware that checks

